### DISCUSSION TITLE
Efficient CI/CD strategies & guidelins for Airflow management in production

### DISCUSSION THREAD

**Discussion by MattiaGallegati** [apache/airflow#30272] [Q&A] [answered]
Hello everyone,
As part of my Airflow adoption, I'm exploring different possibilities for CI/CD and general management of my Airflow installation in production in a complex environment.
I can see that managing lifecycle of DAGs is not an easy task and sure it depends on the context.
Are there/ Can we provide some useful "general" guidelines about these aspects?
There are a lot of grey area where the architect plays an hard role, such as:
- How to manage DAG version and versioning.
- Multi branch/Mono branch approach.
- How and where to test dags during the CI/CD process.
- How to manage the delivery of a new version of an existing DAG while an Airflow instance is running (and maybe also the precedent version of that DAG).
- How to manage environments-dependant configuration in a multi-environment.
And so on..

Do we have something to answer these question?
Do we aim to provide some kind of "general" guidelines?

**Comment by potiuk** [comment, marked as answer]
All of these questions are very general and you need to - basically answer them yourself. Airflow is basically a platform that has plenty of knobs to turn and pieces to put together to make it works - that's both blessing and a curse. Blessing - becasue you can flexibly put several pieces together in exactly the way you want. Curse, because there is no one way to do all the things you mention.

The best you can do you can learn from others: https://airflowsummit.org/ - there are recordings from past 3 years of many users who explained how they are doing some of the stuff you mentioned in "their" environment.  One of my favourit talks is this one https://www.youtube.com/watch?v=uA-8Lj1RNgw where Amum from Jagex explained how they've done that. But they ended up wiht separate git repo for each DAG - effectively having 190 or so git repos put together by submodules. 

Is it good for you when you start - probably not, how your team is organized, which CI you have? Which is the DAG distribution mechanism you choose and deployment you have https://airflow.apache.org/docs/helm-chart/stable/production-guide.html#managing-dag-files ? 

There are many questions - and many answers and I think for most of those you need to simply find the right answer. 

Our docs/prodution guides are pretty comprehensive - but out of necessity, they provide you with the options, and are as little opinionated as possible on the choices - for example you have a description on how you can install Airflow - with pros/cons and expectations you need to have to manage each of the installation method https://airflow.apache.org/docs/apache-airflow/stable/installation/index.html - but the choice is yours. 

Just one thing - speaking of versioning  - complete versionign is not implemented - there is a proposal but it is draft and not even discussed and approved yet https://cwiki.apache.org/confluence/display/AIRFLOW/AIP-36+DAG+Versioning - generally your DAGs should change slowly (and rather in additive fashion, and any "seroius" change in the structure should create a new DAG.  That makes your life a bit more difficult (but also easier, as you know you need to have a new DAG if you change it a lot).

**Reply by MattiaGallegati** [reply]
Thank you,
Very interesting talk, I ended up almost to the same conclusions, but I'm wondering what is the best way to deliver DAG in a prod containerized environement, he actually shows that he is releasing a new Airflow Image for every patch, does this mean the he will stop every DAG that is running and restart the installation at every release? Do you think it's acceptable to simply copy the DAG files or to install them as pip packages on the target machine?

Btw, so we won't provide some kind of guidelines about these aspects and everyone will search their own path.

**Reply by potiuk** [reply]
Sure you can do it as well. I think everyone's mileage is different and what is good for some is not good for others - because of the variety of installatins, team sizes, practices in the company - each Airflow deployment is different and provicding a single solution that would work for everyone is not really possible. Airlfow is a platform and you tune it to your  ways.

> Btw, so we won't provide some kind of guidelines about these aspects and everyone will search their own path.

As you rightfully noticed - you are not comfortable with pushing images with every change - because your case is different. But it worked for Jagex, So **just** this statement of yours underlines that it's YOU who have to make the choice based on analysis what's best for you.
Managed services provide more opinionated solutions - so you might look at them if you need someone to make the choice for you. But if you choose to install it on your own, then only you know what's best for you.

> Do you think it's acceptable to simply copy the DAG files or to install them as pip packages on the target machine?

Sure.. Installing them via pip packages is not realy ok, because you need to make sure they will find their way in DAG folder, and not in system libraries (unless you do some hacks to link them or copy there). You can (and shoulld) install plugins as packages though (see plugins docs).  And as mentioned above - you have some options describe there: https://airflow.apache.org/docs/helm-chart/stable/production-guide.html#managing-dag-files  for example.  But it's not exhaustive list - you should make sure that DAG  find their way into to all the components. 

I am personally big fan of using GitSync for that purpose and have my highly opinionated opinion so I shared it in https://medium.com/apache-airflow/shared-volumes-in-airflow-the-good-the-bad-and-the-ugly-22e9f681afca so you can also read that. But  others have different opinion - I think only you eventually can judge what's best for you, I am afraid,

But if you think you can provide some more guidance that will be generic enough - feel free to submit them. We already have for example this doc that explains how you can install airflow and discusses pros/cons:  https://airflow.apache.org/docs/apache-airflow/stable/installation/index.html  - having simlar explanation for various  DAG distribution methods could be nice, but I am afraid it would have to be similarly "neutral" in terms of explaining the options and describing pros/cons - if you (for example) feel like it - making PRs is always welcome, airflow has around 2.400 contributors and many of them contributed docs as well :)
